pinner/pinner.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * pin.c
+ *
+ * Copyright 2024 Andy Alt <arch_stanton5995@proton.me>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ *
+ */
+
+#include <geanyplugin.h>
+
+static GtkWidget *pinned_view_vbox;
+static gint page_number = 0;
+
+
+static void pin_activate_cb(GtkMenuItem *menuitem, gpointer pdata, GSList *pin_list)
+{
+	GeanyPlugin *plugin = pdata;
+	GeanyDocument *doc = document_get_current();
+
+	if (doc != NULL && pin_list == NULL)
+	{
+		GtkWidget *label = gtk_label_new_with_mnemonic(doc->file_name);
+		gtk_widget_show(label);
+		gtk_box_pack_start(GTK_BOX(pinned_view_vbox), label, FALSE, FALSE, 0);
+		gtk_notebook_set_current_page(GTK_NOTEBOOK(plugin->geany_data->main_widgets->sidebar_notebook), page_number);
+	}
+}
+
+
+static gboolean pin_init(GeanyPlugin *plugin, gpointer pdata)
+{
+	GtkWidget *main_menu_item;
+	GSList *pin_list = NULL;
```

Comment:
How do I pass 'pin_list' to pin_activate_cb()? It seems like I can't just add a 3rd argument, or it's getting ignored. :(

Reply:
Yeah, a third argument doesn't work :(.

The common way is to define a struct to hold the data you want to pass, in this case the plugin and the pin_list pointers and then allocate, fill and pass the struct from `pin_init` to `pin_activate_cb` as the `data` argument.

Care should be paid to the allocation of the struct, it might get tricky to deallocate it in the handler function. I don't remember if GLib offers some utility function for this.
ALternatively, the struct could be created once per plugin and the carried around and reused.

To get a rough idea, see https://github.com/geany/geany-plugins/blob/8431a100f118aacf438af159a4b822469534077f/addons/src/ao_bookmarklist.c#L461.

I didn't test it but maybe you could also use https://www.geany.org/manual/reference/plugindata_8h.html#a143ba8805bd049f3fb13037fae0eb30a to add the `pin_list` to the `plugin` instance. I'm not sure if this is a good idea, it might  get clumsy.

Reply:
Thank you! I'm not good at reading code, but I got the general idea of what needed to be done.